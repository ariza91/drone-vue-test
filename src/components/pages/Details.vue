<template>
  <div class="details">
    <DronePageWrapper :content="wrapper">
      <template slot="title">
        <h3>{{ currentDrone ? currentDrone.name : "Drone details" }}</h3>
      </template>
      <template slot="content" slot-scope="slotProps">
        <div v-if="slotProps.show">
          <div class="averages d-flex justify-content-around mb-5">
            <div
              class="item text-center"
              v-for="(avg, index) in averages"
              :key="index"
            >
              {{ $t(`DroneColumns.${avg.key}`) }}
              <span class="badge bg-primary"> {{ avg | round }}</span>
            </div>
          </div>
          <b-table
            responsive
            hover
            :fields="columns"
            :items="reportFiltered"
            :sort-by="sortOptions.sortBy"
            :sort-desc="sortOptions.sortDesc"
            class="table"
          >
            <template #cell(traffic_conditions)="data">
              <TrafficBadge :text="data.item.traffic_conditions" />
            </template>
          </b-table>
        </div>
      </template>
    </DronePageWrapper>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
import DronePageWrapper from "@/components/modules/DronePageWrapper";
import TrafficBadge from "@/components/elements/TrafficBadge";
import columnsOptions, { sortOptions } from "@/constants/drone-details";
import { calculateAverageOf } from "@/utils/cals";
import { compareAllElementsOf } from "@/utils/filter";

export default {
  name: "DroneDetails",
  components: {
    DronePageWrapper,
    TrafficBadge,
  },
  data() {
    return { sortOptions };
  },
  watch: {
    "$route.params": {
      immediate: true,
      handler(values) {
        this.getDetails(values.id);
      },
    },
  },
  filters: {
    round({ key, value }) {
      return sortOptions.roundValues[key](value);
    },
  },
  computed: {
    ...mapGetters({
      isLoading: "isLoading",
      currentDrone: "currentDrone",
      getSearchInput: "getSearchInput",
    }),
    wrapper() {
      if (!this.currentDrone) return this.$t("DroneDetails.notFound");
      return !this.currentDrone.report.length
        ? this.$t("DroneDetails.notData")
        : null;
    },
    columns() {
      return columnsOptions;
    },
    averages() {
      if (!this.currentDrone) return;
      return calculateAverageOf(this.currentDrone.report, [
        "speed",
        "latitude",
        "longitude",
      ]);
    },
    reportFiltered() {
      return this.currentDrone.report.filter(
        (report) =>
          !this.getSearchInput ||
          compareAllElementsOf(report, this.getSearchInput)
      );
    },
  },
  created() {
    const droneId = this.$route.params.id;
    if (!droneId) return this.$router.push({ name: "Home" });
    this.getDetails(droneId);
  },
  methods: {
    getDetails(droneId) {
      if (!droneId) return;
      this.$store.dispatch("getDroneDetails", droneId);
    },
  },
};
</script>

<style scoped lang="scss">
@import "@/style/theme.scss";

img {
  width: 50%;
}

.table {
  background-color: white;
  border-radius: 5px;
  text-align: center;

  /deep/ th span.sr-only {
    display: none;
  }
}
</style>