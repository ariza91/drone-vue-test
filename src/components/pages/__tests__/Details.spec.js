import { createLocalVue, shallowMount } from "@vue/test-utils";
import Details from "@/components/pages/Details.vue";
import Vuex from 'vuex'

jest.mock('@/store');
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.filter("round", (data) => data);

const mockInput = 'test';
const mockDrone = {
    id: 1,
    batery: 10,
    report: [
        { id: 1, drone_id: 1, batery: mockInput },
        { id: 2, drone_id: 1, batery: 'other' },
    ]
}

const factory = (store) => shallowMount(Details, {
    localVue,
    store,
    mocks: {
        $t: (value) => value,
        $route: { params: { id: 1 } }
    },
    stubs: ['router-link']
});

describe("Details", () => {
    let cmp;
    let getters;
    let actions;
    let store;

    beforeEach(() => {
        getters = {
            currentDrone: () => mockDrone,
            getSearchInput: () => null,
        }
        actions = {
            getDroneDetails: jest.fn()
        };
        store = new Vuex.Store({ getters, actions })
        cmp = factory(store);
        jest.clearAllMocks()
    })
    describe('renders', () => {
        it("renders correctly", () => {
            expect(cmp).toMatchSnapshot()
        });
    })
    describe('watch', () => {
        describe('$route.params', () => {
            it("params change load details", async () => {
                const spy = jest.spyOn(cmp.vm, 'getDetails');
                cmp.vm.$options.watch.$route.call(cmp.vm, { params: { id: 2 } })
                expect(spy).toHaveBeenCalledWith(2)
            });
        });
    })
    describe('computed', () => {
        describe('wrapper', () => {
            it('not current drone list then error', () => {
                getters = { ...getters, currentDrone: () => null }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.wrapper).toEqual("DroneDetails.notFound");
            })
            it('not report list then error', () => {
                getters = {
                    ...getters,
                    currentDrone: () => ({ ...mockDrone, report: [] })
                }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.wrapper).toEqual("DroneDetails.notData");
            })
            it('drone list then null', () => {
                expect(cmp.vm.wrapper).toEqual(null);
            })
        })
        describe('reportFiltered', () => {
            it('not filter then all', () => {
                expect(cmp.vm.reportFiltered).toEqual(mockDrone.report);
            })
            it('filters the reports which contains the input', () => {
                getters = { ...getters, getSearchInput: () => mockInput }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.reportFiltered).toEqual([mockDrone.report[0]]);
            })
            it('filters without reports', () => {
                getters = { ...getters, getSearchInput: () => 'something' }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.reportFiltered).toEqual([]);
            })
        })
    })
    describe('methods', () => {
        describe('methods', () => {
            it('not droneId nothing', () => {
                const spy = jest.spyOn(cmp.vm.$store, 'dispatch');
                cmp.vm.getDetails();
                expect(spy).not.toHaveBeenCalled();
            })
            it('with droneId getDroneDetails called', () => {
                const spy = jest.spyOn(cmp.vm.$store, 'dispatch');
                const droneId = 1;
                cmp.vm.getDetails(droneId);
                expect(spy).toHaveBeenCalledWith("getDroneDetails", droneId);
            })
        });
    })
});
