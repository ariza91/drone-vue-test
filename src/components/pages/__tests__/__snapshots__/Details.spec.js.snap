// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders renders correctly 1`] = `
<div class="details">
  <dronepagewrapper-stub><template></template></dronepagewrapper-stub>
</div>
`;
